keps/0004-test-composability.md

```diff
@@ -0,0 +1,100 @@
+---
+kep-number: 4
+short-desc: This KEP proposes a scheme for making test steps, asserts, and manifests composable in order to prevent repetition of YAML across test cases.
+title: KUTTL Test Composability
+authors:
+  - "@jbarrick-mesosphere"
+owners:
+  - "@jbarrick-mesosphere"
+editor: "@kensipe"
+creation-date: 2020-02-12
+last-updated: 2020-02-25
+status: provisional
+---
+
+# KUTTL Test Composability
+
+This KEP proposes a scheme for making test steps, asserts, and manifests composable in order to prevent repetition of YAML across test cases.
+
+## Table of Contents
+
+* [Summary](#summary)
+  * [Motivation](#motivation)
+     * [Goals](#goals)
+  * [Proposal](#proposal)
+     * [Test Case file structure](#test-case-file-structure)
+
+## Summary
+
+KUTTL tests can tend to be quite repetitive, for example, for test case environment setup or if there are any common resources across test cases. Additionally, sometimes it is useful to compose common asserts.
+
+## Motivation
+
+This is KEP intends to improve the developer experience of writing tests, make tests easier to maintain and reduce duplication by making the files within them more composable. For example, one user [indicated that the tests are overly verbose](https://github.com/kudobuilder/kudo/issues/1311#issuecomment-580709826). Additionally, because this duplication can be expensive, some users deploy common resources into a single namespace breaking namespace isolation and it can make orchestrating tests difficult.
+
+Beyond just composability, the magic file name scheme indicating step index and whether or not a file is an `assert` or `errors` file can be opaque and hard to understand for new users. The proposal in this KEP will also make test structure easier to understand.
+
+### Goals
+
+* Improve composability of test steps, manifests, and asserts.
+* Improve reusability of YAML files.
+* Provide better integration with existing repository structures (e.g., deploy operator YAMLs from a release directory).
+* Make test file structure easier to understand.
+
+## Proposal
+
+The `TestStep` resource will be extended to allow users to provide manifest, errors, and assert files as a list. Three settings will be added: `apply` , `assert`, `error`. Each is a list of files that can either be a single file name or a directory to recurse over.
+
+* `apply`: a list of files to apply for the test step.
+* `assert`: a list of files containing test asserts to expect.
+* `error`: a list of files containing test errors that will cause the test to fail if their state is observed.
+
+An example use of this might look like:
+
+```
+apiVersion: kudo.dev/v1alpha1
+kind: TestStep
+apply:
+- ./install.yaml
+- ../../deploy/
+assert:
+- ../common/check-install.yaml
+error:
+- ./install-error.yaml
+```
+
+This will compose well with the existing `commands` feature of `TestSteps` (note: `commands` will be run prior to either `apply`, `assert`, or `error` during the actual test execution).
```

Comment:
Please also mention the `delete` feature of a `TestStep` and add if this will be run prior to `commands`. Especially for a `TestCase` mentioned below, users may want to clean up deployments at the end of a test case and will use `delete` for that.

Reply:
@nfnt the `delete` feature should be the same as the others and the `commands` would run prior to it... right?   I can't understand if you are wanting to capture the missing `delete`... or if you are expecting different behavior... or both.   Could you explain any justification other than making sure it is included?

Reply:
I'm not sure what the behavior is right now, that's why we should document the expected ordering here. In my opinion, we should run `delete` first, followed by `commands`, then `apply`, `assert`, `error`. Running `delete` first would ensure that the following features would fail when trying to use resources that just have been deleted. Having a different ordering could create subtle bugs.

Reply:
There is great value in documenting and clearing up kuttl test lifecycle.  I don't think it should hold up this kep.  after a lot of thought... It seems like a separate effort than what this kep is trying to accomplish.